### DISCUSSION TITLE
아이템 27. 비검사 경고를 제거하라

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Effective-Java/Effective-Java#55] [3. 과제] [answered]
## 요약

- 비검사 경고는 중요하니 무시하지 말자. 할 수 있는 한 모든 비검사 경고(*Unchecked Warning*)를 제거하자.
  모두 제거한다면 그 코드는 런타임에 `ClassCastException` 이 발생할 일이 없고, 타입 안정성이 보장된다.
- 경고를 제거할 수는 없지만 타입 안전하다고 확신할 수 있다면 `@SuppressWarnings("unchecked")` 으로 경고를 숨기자.
- `@SuppressWarnings("unchecked")` 애노테이션은 항상 가능한 한 좁은 범위에 적용하자.
- `@SuppressWarnings("unchecked")` 애노테이션을 사용할 때면 그 경고를 무시해도 안전한 이유를 항상 주석으로 남겨야 한다.

## 비검사 경고란?

> *Unchecked Warning* 을 번역한 것으로, **컴파일러가 주는 경고 메시지** 정도로 이해하면 된다.

## 비검사 경고의 예시

제네릭에서 새로 생성하는 인스턴스에 타입 매개변수를 명시하지 않으면 IDE 에서는 다음과 같이 표시한다.

![effective-java-item27-1](https://user-images.githubusercontent.com/38161720/211187588-2be2b73f-5c79-49df-a33a-7699727361d6.png)

노란색으로 형광펜이 칠해져 있는 곳에 마우스 커서를 갖다 대면 다음과 같이 안내 메시지를 알려준다.

![effective-java-item27-2](https://user-images.githubusercontent.com/38161720/211187688-303f3b97-3823-44c1-bf88-2d5b9204438f.png)

가능한 모든 비검사 경고를 제거하면 타입 안정성이 보장되므로 이 경우 다이아몬드 연산자(`<>`)를 붙임으로써 해결할 수 있다.
다른 방법으로는 구글에서 개발한 [Guava](https://blog.outsider.ne.kr/710) 라이브러리를 사용할 수 있다.

## 경고를 숨길 수 있는 꼼수

- 자바에서 제공하는 `@SuppressWarnings("unchecked")` 애노테이션을 통해 경고를 숨길 수도 있다.
- 단, 타입 안전성을 검증하지 않은 채 경고만 숨긴 꼴으로, 스스로에게 잘못된 보안 인식을 심어준 셈이다.
- 코드 자체는 경고 없이 컴파일되겠지만, 런타임에서는 여전히 `ClassCastException` 예외가 발생할 수 있다.
- 또한, 안전하다고 검증된 비검사 경고를 숨기지 않고 그대로 두면, 진짜 문제를 알리는 새로운 경고가 나와도 눈치채지 못할 수 있다.

## `@SuppressWarnings` 사용 예시

- 한 줄이 넘는 메서드나 생성자에 달린 `@SuppressWarnings` 애노테이션을 발견하면 지역변수 선언 쪽으로 옮기자.
- 이를 위해 지역변수를 새로 선언해야 하는 수고로움이 있을 수도 있지만, 그만한 값어치가 있다.
- `ArrayList#toArray()` 메서드를 예시로 살펴보면, 현재 메서드 스코프에 `@SuppressWarnings` 가 선언되어 있다.

![effective-java-item27-3](https://user-images.githubusercontent.com/38161720/211188167-d56cbfdb-8aff-4eb4-b1f9-bf4566e6f122.png)

이를 가능한 지역변수 선언 쪽으로 옮긴 후, 경고를 무시해도 안전한 이유를 주석으로 남긴 결과는 다음과 같다.

```java
public <T> T[] toArray(T[] a) {
    if (a.length < size) {
        // 생성한 배열과 매개변수로 받은 배열의 타입이 모두 T[] 로 같으므로 올바른 형변환이다.
        @SuppressWarnings("unchecked")
        T[] result = (T[]) Arrays.copyOf(elementData, size, a.getClass());

        return result;
    }

    System.arraycopy(elementData, 0, a, 0, size);

    if (a.length > size) {
        a[size] = null;
    }

    return a;
}
```

이렇게 주석으로 남기면 다른 사람이 코드를 이해하는 데 도움이 되며, 다른 사람이 코드를 잘못 수정하여 타입 안정성을 잃는 상황을 줄여준다.

**Comment by JoisFe** [comment]
정리 감사합니다 ㅎㅎ

코드를 작성하다 보면 비검사 경고를 확인함에도 당장 귀찮다는 이유로 실행만 잘 되면 제거하지 않고 넘어가는 경우가 있었습니다.
하지만 당장 제거하지 않는다면 프로젝트가 커지고 나서 side effect가 발생하여 애플리케이션에 문제가 일어날 수 있기에 해당 주제를 항상 생각하며 프로젝트를 진행해야 할 것 같습니다.

**Comment by jinan159** [comment, marked as answer]
깔끔한 정리 덕분에 아이템 이해하는데 도움이 많이 됐습니다~!
특히 타입 캐스팅을 할때 해당 경고가 많이 발생하던데, 저도 확실히 타입이 보장되는 상황에서는 사용해야겠네요!
실제 사용사례가 있을까 궁금해서 뒤져보다가 ngrinder 라는 Naver 의 오픈소스에서도 활용하고있는걸 찾아서 공유합니다. 
(https://github.com/naver/ngrinder/blob/develop/ngrinder-core/src/main/java/org/ngrinder/common/util/TypeConvertUtils.java)

```java
package org.ngrinder.common.util;

public abstract class TypeConvertUtils {
	@SuppressWarnings("unchecked")
	public static <T> T cast(Object object) {
		return (T) object;
	}
}
```

![image](https://user-images.githubusercontent.com/45728407/212614598-7cd28632-4b80-4f43-94bd-ac7a092e8700.png)
(cast 메서드 활용)

이미지의 `resolveReference("request")` 는 Object 를 반환하는데, cast 메서드는 반환타입에 맞게 변환해줍니다.
이런 유틸로 캐스팅을 하면 타입 캐스팅 코드가 없어도 되고, Warning 도 모두 제거해주는 장점이 있을 것 같습니다.
대신 남용한다면 타입 안정성을 해칠 수 있으니 말씀하신 것 처럼 확실한 상황에만 사용하는게 좋겠네요..!
